### PR TITLE
nan-only thresholding

### PR DESCRIPTION
**Pull request by adrinjalali**
This adds supporting for thresholding on nans. It surfaces some curious behavior/issue which still needs fixing.

@NicolasHug @ogrisel

### CODE REVIEW

**Comment by NicolasHug** [comment]
> It surfaces some curious behavior/issue which still needs fixing.

can you describe them?

File: sklearn/ensemble/_hist_gradient_boosting/splitting.pyx [hunk 1]
```diff
@@ -406,7 +411,8 @@ cdef class Splitter:
                     feature_idx, histograms, n_samples,
                     sum_gradients, sum_hessians, &split_infos[feature_idx])
 
-                if has_missing_values[feature_idx]:
+                if (has_missing_values[feature_idx]
+                        and not split_infos[feature_idx].split_is_nan):
```
Comment (NicolasHug) [inline sklearn/ensemble/_hist_gradient_boosting/splitting.pyx]: nice!!

File: sklearn/ensemble/_hist_gradient_boosting/splitting.pyx [hunk 2]
```diff
@@ -639,3 +650,21 @@ cdef inline Y_DTYPE_C negative_loss(
...
+cdef inline unsigned char sample_goes_left(
+        unsigned char split_on_nan,
+        unsigned char missing_go_to_left,
+        unsigned char missing_values_bin_idx,
+        X_BINNED_DTYPE_C split_bin_idx,
+        X_BINNED_DTYPE_C bin_value) nogil:
+    return (
+        (
+            split_on_nan and
+            bin_value != missing_values_bin_idx)
```
Comment (NicolasHug) [inline sklearn/ensemble/_hist_gradient_boosting/splitting.pyx]: skip line before `)`?


also maybe add a small comment like

`# if we split on nan, nans always go to right child.`

**Comment by adrinjalali** [comment]
The test I've added passes, but I think the test before that, needs to have `n_bins_non_missing=9`, but it fails with that (that's why I left the comment on your PR)

**Review by NicolasHug — COMMENTED** [review]
I took a quick glance, that looks correct so far.

I wish we could avoid setting the threshold to nan and avoid using `split_is_nan`. I don't remember seeing those in lightgbm.

About the use of "nan" vs "missing values": in the code I chose to almost never use "nan" and use instead "missing values", since missing values can potentially be represented by something other than a nan. E.g. zero entries in a  sparse matrix. I'm OK with using nan anyway since that's the only supported use-case right now.

File: sklearn/ensemble/_hist_gradient_boosting/splitting.pyx (hunk 1 above)
Comment (adrinjalali) [reply sklearn/ensemble/_hist_gradient_boosting/splitting.pyx]: yeah this was @ogrisel!

**Comment by NicolasHug** [comment]
can you add a test for infinite values as input and see if we can still support them when we set the threshold to +- inf (hope it doesn't add too much code)

**Comment by NicolasHug** [comment]
If we "split on nan", missing values go to the right node.

It is slightly confusing that a sample with missing value may go to the right child, despite `missing_go_to_left` being True. Ideally, only `missing_go_to_left` should decide where these samples go. We could get rid of the `sample_goes_left` function.

I think that this can be solved by scanning from right to left first, instead of the current version which is scanning from left to right first.

**Comment by NicolasHug** [comment]
This test passes on the PR branch but fails here:

```py
        ([8, 0, 1, 8, 2, 3, 4, 5, 6, 7],  # 8 <=> missing
         [1, 1, 1, 1, 5, 5, 5, 5, 5, 5],
         True,  # missing values
         8,  # n_bins_non_missing
         1,  # cut on bin_idx=1
         True),  # missing values go to left
```

Note that the missing bin is 8 here (not 9).

That seems like a regression

File: sklearn/ensemble/_hist_gradient_boosting/splitting.pyx [hunk 3]
```diff
@@ -556,7 +565,7 @@ cdef class Splitter:
             Y_DTYPE_C sum_gradient_right
             Y_DTYPE_C negative_loss_current_node
             Y_DTYPE_C gain
-            unsigned int start = self.n_bins_non_missing[feature_idx] - 2
+            unsigned int start = self.n_bins_non_missing[feature_idx] - 1
```
Comment (NicolasHug) [inline sklearn/ensemble/_hist_gradient_boosting/splitting.pyx]: By using `-1`, you are counting the missing values bin as part of the right node at the first iteration of the loop.

I still need to convince myself that this is the right fix (and that we don't need something else too), but reverting it to `-2` seems to fix the problem.

**Comment by NicolasHug** [comment]
~For some reason I can't push to your branch.~ (OK I figured it out)

I fixed the previous issues and added a few more tests. I hope they make more sense and avoid confusion. Here's the patch.

```patch
diff --git a/sklearn/ensemble/_hist_gradient_boosting/splitting.pyx b/sklearn/ensemble/_hist_gradient_boosting/splitting.pyx
index f444248ae..b08b580e6 100644
--- a/sklearn/ensemble/_hist_gradient_boosting/splitting.pyx
+++ b/sklearn/ensemble/_hist_gradient_boosting/splitting.pyx
@@ -565,7 +565,7 @@ cdef class Splitter:
             Y_DTYPE_C sum_gradient_right
             Y_DTYPE_C negative_loss_current_node
             Y_DTYPE_C gain
-            unsigned int start = self.n_bins_non_missing[feature_idx] - 1
+            unsigned int start = self.n_bins_non_missing[feature_idx] - 2
 
         sum_gradient_right, sum_hessian_right = 0., 0.
         n_samples_right = 0
diff --git a/sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py b/sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py
index 859c6ccdb..77bae28a4 100644
--- a/sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py
+++ b/sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py
@@ -286,10 +286,18 @@ def test_min_gain_to_split():
          3,  # expected_bin_idx
          'not_applicable'),
 
-        # We replace 2 samples by NaNs (bin_idx=9)
+        # We replace 2 samples by NaNs (bin_idx=8)
         # These 2 samples were mapped to the left node before, so they should
         # be mapped to left node again
         # Notice how the bin_idx threshold changes from 3 to 1.
+        ([8, 0, 1, 8, 2, 3, 4, 5, 6, 7],  # 8 <=> missing
+         [1, 1, 1, 1, 5, 5, 5, 5, 5, 5],
+         True,  # missing values
+         8,  # n_bins_non_missing
+         1,  # cut on bin_idx=1
+         True),  # missing values go to left
+
+        # same as above, but with non-consecutive missing_values_bin
         ([9, 0, 1, 9, 2, 3, 4, 5, 6, 7],  # 9 <=> missing
          [1, 1, 1, 1, 5, 5, 5, 5, 5, 5],
          True,  # missing values
@@ -297,7 +305,15 @@ def test_min_gain_to_split():
          1,  # cut on bin_idx=1
          True),  # missing values go to left
 
-        # Same, this time replacing 2 samples that were on the right.
+        # this time replacing 2 samples that were on the right.
+        ([0, 1, 2, 3, 8, 4, 8, 5, 6, 7],  # 8 <=> missing
+         [1, 1, 1, 1, 5, 5, 5, 5, 5, 5],
+         True,  # missing values
+         8,  # n_bins_non_missing
+         3,  # cut on bin_idx=3 (like in first case)
+         False),  # missing values go to right
+
+        # same as above, but with non-consecutive missing_values_bin
         ([0, 1, 2, 3, 9, 4, 9, 5, 6, 7],  # 9 <=> missing
          [1, 1, 1, 1, 5, 5, 5, 5, 5, 5],
          True,  # missing values
@@ -368,21 +384,40 @@ def test_splitting_missing_values(X_binned, all_gradients,
     'X_binned, all_gradients, has_missing_values, n_bins_non_missing, '
     ' expected_bin_idx, split_is_nan, expected_go_to_left', [
 
-        ([0, 1, 2, 3, 7, 8, 9, 9, 9, 9],  # 9 <=> missing
+        ([0, 1, 2, 3, 4, 5, 6, 6, 6, 6],  # 6 <=> missing
          [1, 1, 1, 1, 1, 1, 5, 5, 5, 5],
          True,  # missing values
-         9,  # n_bins_non_missing
-         8,  # cut on bin_idx=max_bins - 1
+         6,  # n_bins_non_missing
+         5,  # cut on bin_idx=max_bins - 1
          True,  # split_is_nan
          False),  # missing values go to right
 
-        ([9, 9, 9, 9, 9, 9, 1, 3, 8, 6],  # 9 <=> missing
+        # same as above, but with non-consecutive missing_values_bin
+        ([0, 1, 2, 3, 4, 5, 9, 9, 9, 9],  # 9 <=> missing
          [1, 1, 1, 1, 1, 1, 5, 5, 5, 5],
          True,  # missing values
-         9,  # n_bins_non_missing
-         8,  # cut on bin_idx=max_bins - 1
+         6,  # n_bins_non_missing
+         5,
          True,  # split_is_nan
          False),  # missing values go to right
+
+        ([4, 4, 4, 4, 4, 4, 0, 1, 2, 3],  # 4 <=> missing
+         [1, 1, 1, 1, 1, 1, 5, 5, 5, 5],
+         True,  # missing values
+         4,  # n_bins_non_missing
+         3,  # cut on bin_idx=max_bins - 1
+         True,  # split_is_nan
+         False),  # missing values go to right
+
+        # same as above, but with non-consecutive missing_values_bin
+        ([9, 9, 9, 9, 9, 9, 0, 1, 2, 3],  # 9 <=> missing
+         [1, 1, 1, 1, 1, 1, 5, 5, 5, 5],
+         True,  # missing values
+         4,  # n_bins_non_missing
+         3,  # cut on bin_idx=max_bins - 1
+         True,  # split_is_nan
+         False),  # missing values go to right
+
     ]
 )
 def test_splitting_missing_values_edge_case(
@@ -433,5 +468,6 @@ def test_splitting_missing_values_edge_case(
     samples_left, samples_right, _ = splitter.split_indices(
         split_info, splitter.partition)
 
-    nan_idx = np.flatnonzero(np.array(X_binned) == n_bins_non_missing)
-    assert set(samples_right) == set(nan_idx)
+    missing_samples_indices = np.flatnonzero(
+        np.array(X_binned) == missing_values_bin_idx)
+    assert set(samples_right) == set(missing_samples_indices)
```

**Comment by adrinjalali** [comment]
Awesome, thanks @NicolasHug! @ogrisel had some ideas for some other tests, but I guess we can merge this into your branch now, and continue the discussion on the main PR?

**Comment by NicolasHug** [comment]
Cool yeah I'll merge

**Pull request merged**